src/main/java/org/apache/flink/benchmark/SortingBoundedInputBenchmarks.java

```diff
@@ -0,0 +1,418 @@
+import org.apache.flink.streaming.api.operators.StreamOperatorFactory;
+import org.apache.flink.streaming.api.operators.StreamOperatorParameters;
+import org.apache.flink.streaming.api.operators.TwoInputStreamOperator;
+import org.apache.flink.streaming.api.transformations.KeyedMultipleInputTransformation;
+import org.apache.flink.streaming.runtime.streamrecord.LatencyMarker;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.util.SplittableIterator;
+
+import org.junit.Assert;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.OperationsPerInvocation;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+import org.openjdk.jmh.runner.options.VerboseMode;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Random;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import static org.openjdk.jmh.annotations.Scope.Thread;
+
+/**
+ * An end to end test for sorted inputs for a keyed operator with bounded inputs.
+ */
+public class SortingBoundedInputBenchmarks extends BenchmarkBase {
+
+	protected static final int RECORDS_PER_INVOCATION = 900_000;
+
+	public static void main(String[] args) throws RunnerException {
+		Options options = new OptionsBuilder()
+				.verbosity(VerboseMode.NORMAL)
+				.include(".*" + SortingBoundedInputBenchmarks.class.getCanonicalName() + ".*")
+				.build();
+
+		new Runner(options).run();
+	}
+
+	@State(Thread)
+	public static class SortingInputContext extends FlinkEnvironmentContext {
+		@Param({"true", "false"})
+		public boolean useSingleKeyStateBackend;
+
+		@Override
+		protected Configuration createConfiguration() {
+			Configuration configuration = super.createConfiguration();
+			configuration.set(ExecutionOptions.USE_BATCH_STATE_BACKEND, useSingleKeyStateBackend);
+			configuration.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.BATCH);
+			return configuration;
+		}
+	}
+
+	@Benchmark
+	@OperationsPerInvocation(value = RECORDS_PER_INVOCATION)
+	public void oneInputOperator(SortingInputContext context) {
+		StreamExecutionEnvironment env = context.env;
+
+		DataStreamSource<Tuple2<Integer, byte[]>> elements = env.fromParallelCollection(
+			new InputGenerator(RECORDS_PER_INVOCATION),
+			new TupleTypeInfo<>(BasicTypeInfo.INT_TYPE_INFO, PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO)
+		);
+
+		SingleOutputStreamOperator<Long> counts = elements
+			.keyBy(element -> element.f0)
+			.transform(
+				"Asserting operator",
+				BasicTypeInfo.LONG_TYPE_INFO,
+				new AssertingOperator()
+			);
+
+		counts.addSink(new DiscardingSink<>());
```

Comment:
aren't we missing `env.execute()`? The benchmark results I observed are suspiciously high
```
Iteration   1: 125200.947 ops/ms
Iteration   2: 164695.613 ops/ms
Iteration   3: 133524.562 ops/ms
```
This can also explain resource leaks and stability problems.